swap pycryptodome to the faster, smaller, and industry standard cryto…

lithiumFlower opened the pull request:
There is an outstanding issue https://github.com/pdfminer/pdfminer.six/issues/429

While the size of pycryptodome is not as much as titled, it is still quite large (~40 MB).

Cryptography and all of its dependencies is ~7 MB. The dependencies (cffi, six, pycparser) are well known and common in other packages.

Cryptography defaults to using its shipped openssl library and has become the industry standard for python. 
Why is openssl advantageous compared to pycryptodome?
1. All crypto routines are purely written in C and wrapped in python. 
2. openssl utilizes hardware accelerated AES. This is the most significant performance boost a symmetric encryption routine can get.
3. While openssl has had its security shortcomings in the past, it's one of the most used and updated cryto libs in existence. This means comparatively hardened security compared to pycryptodome. Of course, as far as I can tell from my cursory look at the codebase pdfminer only decrypts. This makes the security impact low.

**How Has This Been Tested?**
I ran the shipped test suits including `make crypts` on the sample pdfs.

**Checklist**

- [x] I have added tests that prove my fix is effective or that my feature 
  works (I have not added existing codepaths, the existing testcases test these)
- [x] I have added docstrings to newly created methods and classes (no new methods or classes)
- [x] I have optimized the code at least one time after creating the initial 
  version
- [x] I have updated the [README.md](../README.md) or I am verified that this
  is not necessary
- [x] I have updated the [readthedocs](../docs/source) documentation or I 
  verified that this is not necessary
- [x] I have added a consice human-readable description of the change to 
  [CHANGELOG.md](../CHANGELOG.md)

lithiumFlower commented:
Unfortunately, cryptography supports Python 2.7, Python 3.5+, and PyPy 5.4+,  but not Python 3.4.

pietermarsman commented on setup.py:
```diff
@@ -14,7 +14,7 @@
     install_requires=[
         'chardet ; python_version > "3.0"',
-        'pycryptodome',
+        'cryptography',
```
We could also move cryptography to extra_requires. Then most people wont install cryptography at all. For pycryptodome that made sense, since it has these dependencies on c.

But for cryptography its much less relevant. It only has a few dependencies:

```
cryptography==2.9.2
  - cffi [required: >=1.8,!=1.11.3, installed: 1.14.0]
    - pycparser [required: Any, installed: 2.20]
  - six [required: >=1.4.1, installed: 1.14.0]
```

lithiumFlower replied:
So you do or do not want it moved to `extra_requires`?

corneliusroemer replied:
The issue/request I opened is agnostic about methods of reducing install size. What matters is to get it down for Alpine Linux. I'll see whether I can test it quickly on my docker build.

corneliusroemer replied:
Ok, I'm not good enough with Docker to build the work in progress within Docker and test for size. If you have a neat package I can simply download and install from within I could do the test, though. Otherwise, I guess no harm in implementing it first without `extra_requires` so as to be non-breaking, and then do that in a second step.
I just learned I can also simply uninstall the build tools I had to install again to make the image leaner.

lithiumFlower replied:
I made a comment on the referenced issue about cryptography which builds a wheel for alpine linux.  You should be able to verify for yourself that pdfminer + cryptography will work by simply installing cryptography.

corneliusroemer replied:
Fantastic :) That sounds like switching to cryptography should address the reason for creating the issue.

pietermarsman replied:
Leave it in install_requires. Using extra_requires was a sensible idea if we were stuck with pycryptodome.

pietermarsman commented on pdfminer/pdfdocument.py:
```diff
@@ -485,27 +499,35 @@ def get_cfm(self, name):
     def decrypt_aes256(self, objid, genno, data):
-        return AES.new(self.key, mode=AES.MODE_CBC, IV=data[:16])\
-            .decrypt(data[16:])
+        iv = data[:16]
+        ct = data[16:]
```
Giving these parts a name, but `iv` and `ct` don't really help to understand what it is. You should use a more verbose name.

pietermarsman commented on pdfminer/pdfdocument.py:
```diff
@@ -4,11 +4,20 @@
-    from . import arcfour as ARC4
+    AES = None
+
+try:
+    from hashlib import sha256
+except ImportError:
+    sha256 = None
```
This try-except is not needed anymore

lithiumFlower replied:
I agree. I tried to change it line for line as little as possible as that tends to make many maintainers happy. Will remove

lithiumFlower replied:
Yep, will remove. I tried to change it line for line as little as possible because that makes many library maintainers happy

pietermarsman commented on pdfminer/pdfdocument.py:
```diff
@@ -4,11 +4,20 @@
-    from Crypto.Cipher import ARC4, AES
-    from Crypto.Hash import SHA256
+    from cryptography.hazmat.primitives.ciphers import (Cipher, algorithms,
+                                                        modes)
+    from cryptography.hazmat.backends import default_backend
+    AES = True
 except ImportError:
-    AES = SHA256 = None
-    from . import arcfour as ARC4
+    AES = None
```
This try-except is not needed anymore. I think installing cryptography is no problem for everyone.

lithiumFlower commented:
@pietermarsman When you say that the files in `samples_files/encryption` are not part of the test suite, do you mean they're not used on travis? They are currently run as part of the Makefile's `crypts` task. If you use the parent Makefile and run `make tests` it will run nosetests, the processing of the free and nonfree pdfs in sample_files, as well as the `crypts` target mentioned previously. 

I can add/refactor to use any of these targets instead of the existing tox.ini logic if that's what you're suggesting. However, if travis doesn't have the required programs installed then I'm not interested in debugging that.

The options would be: 
    1. simply call the parent Makefile's test target instead of `nosetests ..` in `tox.init`
    2 leave the current nose call and add an additional child Makefile call to `make crypts`

If that is what you were suggesting then please let me know which would be preferred. If not, then I didn't understand your comment.

corneliusroemer commented:
> There is an outstanding issue #429
> 
> While the size of pycryptodome is not as much as titled, it is still quite large (~40 MB).
> 
> Cryptography and all of its dependencies is ~7 MB. The dependencies (cffi, six, pycparser) are well known and common in other packages.

I'm the one who opened the issue. I think there's a slight misunderstanding as to the origin of the extra 260MB for pycryptodome.

The fact is, that in order to install pdfminer on Alpine Linux, all the requirements can add to 260 MB. I needed to install the following, which all count towards the 260 MB:
`apk add gcc g++ make libffi-dev openssl-dev`

My personal request was to somehow change pdfminer so that it has less bloating (be it pdfminer module proper or requirements) when installing on Alpine Linux. Whether switching to a different crypto package helps, I don't know. It depends as much on its requirements as on the package itself. So I guess one would have to test that specifically.

lithiumFlower commented:
@corneliusroemer w.r.t to your comment https://github.com/pdfminer/pdfminer.six/pull/456#issuecomment-657543648

> I'm the one who opened the issue. I think there's a slight misunderstanding as to the origin of the extra 260MB for pycryptodome.

Thanks for the heads up, though I am aware of those particulars (I commented on the issue itself). If there were a prebuilt wheel for Alpine it would probably be of size similar to the windows size (~40 MB).

corneliusroemer commented:
@lithiumFlower Thanks! That makes sense. Ah right, it's called wheel if you don't need build tools!

pietermarsman commented:
> @pietermarsman When you say that the files in `samples_files/encryption` are not part of the test suite, do you mean they're not used on travis? 

Yes, that's what I mean. My preferred solution is to add them to the nosetest test suite. In particular, `test_pdf2txt.py` is the place to test if we can extract a pdf without error. For the encrypted pdf's we don't have any nosetests yet. I'm also ok with removing the makefile in the encryption directory to eliminate this alternative testing option.

lithiumFlower commented:
@pietermarsman I've removed the samples Makefile and ported its functionality to `tests/test_samples.py`. I've used a new test file as its purpose is different from `test_pdf2txt.py`. Additionally, there were no `.refs` in the repository meaning there was nothing to test against to ensure you've not changed the output, so I added them for each pdf. From the travis build we can see that `nonfree/kamp.pdf` parses differently with python 3.8 than the other versions.

lithiumFlower commented:
@pietermarsman Ok. In the interest of merging the specific improvement of pycryptodome -> cryptography I'm going to revert to before the Makefile port and make that a separate MR that people can do with what they please. Then, making the swap is available and ready if a maintainer chooses to use it.

pietermarsman commented:
Probably easier if I do it myself. Can you give it one more look before I merge it?

lithiumFlower commented:
LGTM

pietermarsman commented:
Great! Thanks for all the work :D
